---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# AnnotateCell

<!-- badges: start -->
<!-- badges: end -->

Cell type identification remains a key challenge in single-cell RNA-seq and spatial transcriptomics datasets despite a decade of research in this area.

There are many reference-based annotation tools to help with cell type annotation. However, the input and output formats differ greatly, increasing the burden on users to use and compare multiple tools. 

This package `AnnotateCell` aims to provide a unified wrapper to run several annotation tools conveniently.


## Installation

You can install AnnotateCell like so:

```{r, eval=FALSE}
library(remotes)
install_github("adairama/AnnotateCell")
```

# Illustration

## Demo dataset

We will load the `pbmc.demo` dataset. This is a Seurat object containing 2,635 peripheral blood mononuclear cells (PBMC) from 10X Genomics experiment. This is the same dataset used in the fundamental Seurat vignette after QC filtering with PCA and UMAP embedding. See help(pbmc.demo) for more details.

```{r dataset}
library(AnnotateCell)

pbmc.demo

pbmc.demo@meta.data %>% head()
```
The authors of Seurat package provided their annotation which is stored as `seurat_annotations` in the meta data.

```{r}
pbmc.demo@meta.data %>% 
  janitor::tabyl(seurat_annotations)
```

## Run RCAv2 annotation

Here is an example on how to run the RCAv2 tool with the GlobalPanel_CellTypes panel.

```{r example}
pred <- AnnotateCell(pbmc.demo, "RCAv2.GlobalPanel_CellTypes")

head(pred)
```

## Comparing the annotation

The first task is to add the prediction to the Seurat object.

```{r}
pbmc.demo <- AddMetaData(pbmc.demo, pred)
```

```{r}
cts <- pbmc.demo@meta.data %>% 
  count(group = seurat_annotations, pred = ann.RCAv2.GlobalPanel_CellTypes) %>% 
  arrange(group, desc(n))
```

