---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)

options(width = 150)
```

# AnnotateCells

Cell type identification remains a key challenge in single-cell RNA-seq and spatial transcriptomics datasets despite a decade of research in this area.

There are many reference-based annotation tools to help with cell type annotation. However, the input and output formats differ greatly, increasing the burden on users to use and compare multiple tools. 

This package `AnnotateCells` aims to provide a unified wrapper to run several annotation tools conveniently.


# Installation

We can ensure all the necessary prerequisites are met via

```{r, eval=FALSE}
if (!requireNamespace("pacman", quietly = TRUE)) install.packages("pacman")

pacman::p_load( Seurat, SingleR, celldex )

pacman::p_load_gh("prabhakarlab/RCAv2")
pacman::p_load_gh("JinmiaoChenLab/DISCOtoolkit")
pacman::p_load_gh("satijalab/Azimuth")
```

Next, you can install AnnotateCells:

```{r, eval=FALSE}
pacman::p_load_gh("adairama/AnnotateCells")
```


# Demo dataset

We will load the `pbmc.demo` dataset. This is a Seurat object containing 2,635 peripheral blood mononuclear cells (PBMC) from 10X Genomics experiment. This is the same dataset used in the fundamental Seurat vignette after QC filtering with PCA and UMAP embedding. See `help(pbmc.demo)` for more details.

```{r dataset}
library(AnnotateCells)

data(pbmc.demo)

pbmc.demo
```

The Seurat package includes a column called `seurat_annotations`, which we will treat as the ground truth for evaluating annotation accuracy.

For the purpose of demonstrating the cell annotation workflow, we will use annotate the 11 clusters found in the `RNA_snn_res.0.8` column. These clusters were generated using the Louvain algorithm with a resolution of 0.8.

```{r}
pbmc.demo@meta.data %>% head()

table(pbmc.demo$RNA_snn_res.0.8)
```

# Available panels

dt


# Run RCAv2 annotation

Here is an example on how to run the RCAv2 tool with the GlobalPanel_CellTypes panel.

```{r}
pred <- AnnotateCells(pbmc.demo, "RCAv2.GlobalPanel_CellTypes")

dim(pred)

head(pred)
```

Let us add the prediction to the Seurat object.

```{r}
pbmc.demo <- AddMetaData(pbmc.demo, pred)
```


# Run DISCO annotation

The first time you run the DISCO annotation, the required reference data (~175 MB) will be downloaded from the server. This may take a few minutes (<5 min), depending on your internet speed. The downloaded data will be stored in the data directory of the `AnnotateCells` package for future use.

```{r}
pred <- AnnotateCells(pbmc.demo, "DISCO.all")

dim(pred)

head(pred)
```


## Aligning predicted cell type label with cluster group

With the exception of the DISCO reference tool, `AnnotateCells()` provides predictions at the individual cell level. Aligning these predicted cell type labels to cluster groups can be advantageous.
However, each cluster may contain multiple prediction labels as the cross table below demonstrates, 

```{r}
table(pbmc.demo$ann.RCAv2.GlobalPanel_CellTypes, pbmc.demo$RNA_snn_res.0.8) %>% 
  print(zero.print = ".")
```

It can be easier to visualize this information (e.g. cluster 3) via the following barplot. 
The segments are ordered in descending order of popularity until at least 70% of the cells are accounted for (you can adjust this using the `thres` parameter). The remaining labels are lumped together as Misc. and colored in grey.

```{r, eval = FALSE}
align_prediction_to_cluster(
  prediction = pbmc.demo$ann.RCAv2.GlobalPanel_CellTypes,
  cluster    = pbmc.demo$RNA_snn_res.0.8 
)
```

```{r, echo = FALSE}
knitr::include_graphics("man/figures/align_prediction_to_cluster.png")
```

Here is an alternative text output:

```{r, eval=FALSE}
align_prediction_to_cluster(
  prediction = pbmc.demo$ann.RCAv2.GlobalPanel_CellTypes,
  cluster    = pbmc.demo$RNA_snn_res.0.8,
  type       = "split"
)
```
